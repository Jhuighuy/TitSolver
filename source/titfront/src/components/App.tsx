/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ *\
 * Part of BlueTit Solver, licensed under Apache 2.0 with Commons Clause.
 * Commercial use, including SaaS, requires a separate license, see /LICENSE.md
\* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

import type { FC } from "react";
import { AiOutlinePython as PythonIcon } from "react-icons/ai";
import {
  FiActivity as ActivityIcon,
  FiDatabase as DatabaseIcon,
  FiSettings as SettingsIcon,
  FiSliders as SlidersIcon,
  FiTerminal as TerminalIcon,
} from "react-icons/fi";

import { Menu } from "~/components/Menu";
import { Viewer } from "~/components/Viewer";

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

export const App: FC = () => {
  const leftIconSize = 24;
  const bottomIconSize = 16;
  return (
    <div className="h-screen w-screen flex flex-row select-none text-sm">
      <Menu side="left">
        <Menu.Item
          name="Configuration"
          icon={<SlidersIcon size={leftIconSize} />}
          group={0}
        />
        <Menu.Item
          name="Storage"
          icon={<DatabaseIcon size={leftIconSize} />}
          group={0}
        />
        <Menu.Item
          name="Activity"
          icon={<ActivityIcon size={leftIconSize} />}
          group={1}
        />
        <Menu.Item
          name="Settings"
          icon={<SettingsIcon size={leftIconSize} />}
          group={1}
        />
      </Menu>
      <div className="flex-1 flex flex-col">
        <Viewer />
        <Menu side="bottom">
          <Menu.Item
            name="Python shell"
            icon={<PythonIcon size={bottomIconSize} />}
            group={0}
          />
          <Menu.Item
            name="Console"
            icon={<TerminalIcon size={bottomIconSize} />}
            group={0}
          />
        </Menu>
      </div>
    </div>
  );
};

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
