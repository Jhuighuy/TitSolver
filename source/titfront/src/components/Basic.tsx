/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ *\
 * Part of BlueTit Solver, under the MIT License.
 * See /LICENSE.md for license information. SPDX-License-Identifier: MIT
\* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

import { Text } from "@radix-ui/themes";
import type { ComponentProps } from "react";

import { cn } from "~/utils";

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

type TechnicalTextProps = ComponentProps<typeof Text> & {
  className?: string;
};

export function TechnicalText({
  className,
  ...props
}: Readonly<TechnicalTextProps>) {
  return <Text {...props} className={cn("font-mono italic", className)} />;
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
