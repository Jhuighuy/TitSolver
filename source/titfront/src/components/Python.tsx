/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ *\
 * Part of the Tit Solver project, under the MIT License.
 * See /LICENSE.md for license information. SPDX-License-Identifier: MIT
\* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

import {
  FC,
  ReactNode,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";
import { z } from "zod";

import { assert } from "../checks";

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/**
 * Python error.
 */
export class PyError extends Error {
  /**
   * Construct a new Python error.
   */
  constructor(
    public readonly type: string,
    public readonly error: string,
    public readonly traceback?: string
  ) {
    super(error);
  }

  /**
   * Get the error message.
   */
  toString() {
    let result = `${this.type}: ${this.error}`;
    if (this.traceback) result += `\n\n${this.traceback}`;
    return result;
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/**
 * Callback to pass the execution result.
 * @param result The result of the Python code or an error.
 */
export type PyCallback = (result: unknown) => void;

/**
 * Run Python code.
 * @param expression The Python code to run.
 * @param onResponse The callback to run when the code finishes.
 */
export type PyRunCode = (expression: string, onResponse: PyCallback) => void;

/**
 * Access connection to the Python server.
 * @returns A function to run Python code.
 */
export function usePython(): PyRunCode {
  const runCode = useContext(PyConnectionContext);
  assert(runCode !== null, "runCode is null!");
  return runCode;
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

const PyConnectionContext = createContext<PyRunCode | null>(null);
const PyResponseSchema = z.union([
  z.object({
    requestID: z.string(),
    status: z.literal("error"),
    result: z.object({
      type: z.string(),
      error: z.string(),
      traceback: z.string().optional(),
    }),
  }),
  z.object({
    requestID: z.string(),
    status: z.literal("success"),
    result: z.unknown(),
  }),
]);

/**
 * Provide a connection to the Python server to the children.
 */
export const PyConnectionProvider: FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [webSocket, setWebSocket] = useState<WebSocket | null>(null);
  const pendingRequests = useRef(new Map<string, PyCallback>());

  useEffect(() => {
    const ws = new WebSocket(`ws://${window.location.host}/ws`);
    ws.onopen = () => setWebSocket(ws);
    ws.onclose = () => setWebSocket(null);
    ws.onmessage = (event: MessageEvent) => {
      // Parse and validate the message.
      assert(typeof event.data === "string", "Not a string message.");
      const { requestID, status, result } = PyResponseSchema.parse(
        JSON.parse(event.data)
      );

      // Invoke the callback.
      const callback = pendingRequests.current.get(requestID);
      assert(callback !== undefined, `No callback for request ${requestID}`);
      if (status === "success") {
        callback(result);
      } else {
        assert(status === "error");
        const { type, error, traceback } = result;
        callback(new PyError(type, error, traceback));
      }
      pendingRequests.current.delete(requestID);
    };

    return () => {
      if (ws.readyState === WebSocket.OPEN) ws.close();
    };
  }, []);

  const runCode = useCallback(
    (expression: string, onResponse: PyCallback) => {
      assert(webSocket !== null, "WebSocket is not null!");
      assert(
        webSocket.readyState === WebSocket.OPEN,
        "Socket is not connected!"
      );

      const requestID = Math.random().toString(36).substring(8);
      pendingRequests.current.set(requestID, onResponse);
      webSocket.send(JSON.stringify({ requestID, expression }));
    },
    [webSocket]
  );

  if (!webSocket) return <div>Connecting...</div>;
  return (
    <PyConnectionContext.Provider value={runCode}>
      {children}
    </PyConnectionContext.Provider>
  );
};

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
