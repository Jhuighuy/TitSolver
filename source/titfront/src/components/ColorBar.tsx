/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ *\
 * Part of BlueTit Solver, under the MIT License.
 * See /LICENSE.md for license information. SPDX-License-Identifier: MIT
\* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

import { Box } from "@radix-ui/themes";
import type { ComponentProps, CSSProperties } from "react";

import { TechnicalText } from "~/components/Basic";
import { iota, type Orientation } from "~/utils";
import { type ColorMapName, colorMaps } from "~/visual/ColorMap";

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

type ColorBoxProps = ComponentProps<typeof Box> & {
  name: ColorMapName;
  orientation?: Orientation;
  style?: CSSProperties;
};

export function ColorBox({
  name,
  orientation = "horizontal",
  ...props
}: ColorBoxProps) {
  const gradientStops = colorMaps[name].points
    .map(
      ([v, r, g, b]) =>
        `rgb(
          ${Math.round(r * 255)},
          ${Math.round(g * 255)},
          ${Math.round(b * 255)}
        ) ${v * 100}%`
    )
    .join(", ");

  const gradient =
    orientation === "vertical"
      ? `linear-gradient(to top, ${gradientStops})`
      : `linear-gradient(to right, ${gradientStops})`;

  return <Box {...props} style={{ ...props.style, background: gradient }} />;
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

type ColorLegendProps = ColorBoxProps & {
  min: number;
  max: number;
  ticks: number;
  title?: string;
};

export function ColorLegendProps({
  min,
  max,
  ticks,
  title,
  orientation = "horizontal",
  ...props
}: ColorLegendProps) {
  const reversed = orientation === "vertical";

  return (
    <ColorBox {...props} orientation={orientation} position="relative">
      {/* ---- Title. ------------------------------------------------------ */}
      {title && (
        <Box
          position="absolute"
          {...(orientation === "vertical"
            ? {
                left: "-100%",
                top: "50%",
                style: { transform: "translate(-50%, -50%) rotate(-90deg)" },
              }
            : {
                left: "50%",
                top: "-100%",
                style: { transform: "translate(-50%, -50%)" },
              })}
        >
          <TechnicalText color="gray">{title}</TechnicalText>
        </Box>
      )}

      {/* ---- Ticks. ------------------------------------------------------ */}
      {iota(ticks).map((index) => {
        const t = index / (ticks - 1);
        const value = min + (max - min) * t;
        const offset = `calc(${reversed ? 1 - t : t} * (100% - 2px) + 1px)`;

        return (
          <Box
            key={index}
            position="absolute"
            {...(orientation === "vertical"
              ? { left: "100%", top: offset, width: "100%", height: "2px" }
              : { left: offset, top: "100%", width: "2px", height: "100%" })}
            style={{
              background: "var(--gray-11)",
              transform: "translate(-50%, -50%)",
            }}
          >
            <Box
              position="absolute"
              {...(orientation === "vertical"
                ? { left: "275%" }
                : { left: "50%", top: "150%" })}
              style={{ transform: "translate(-50%, -50%)" }}
            >
              <TechnicalText size="2" color="gray">
                {value.toFixed(2)}
              </TechnicalText>
            </Box>
          </Box>
        );
      })}
    </ColorBox>
  );
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
