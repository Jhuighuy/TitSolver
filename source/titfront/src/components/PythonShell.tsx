/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ *\
 * Part of BlueTit Solver, licensed under Apache 2.0 with Commons Clause.
 * Commercial use, including SaaS, requires a separate license, see /LICENSE.md
\* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

import {
  FC,
  Fragment,
  KeyboardEvent,
  MouseEvent,
  useEffect,
  useRef,
} from "react";
import {
  BiRadioCircle as CircleIcon,
  BiRadioCircleMarked as FilledCircleIcon,
} from "react-icons/bi";
import {
  FiCornerDownRight as EnterIcon,
  FiDollarSign as DollarIcon,
} from "react-icons/fi";

import { PyError, usePython } from "~/components/Python";
import { usePythonShellStore } from "~/stores/PythonShellStore";
import { cn } from "~/utils";

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/** @todo Implement a timeout for commands. */
export const PythonShell: FC = () => {
  const {
    commands,
    pushCommand,
    updateCommand,
    firstVisibleCommand,
    clearVisibleCommands,
    currentInput,
    setCurrentInput,
    historyIndex,
    historyUp,
    historyDown,
  } = usePythonShellStore();
  const inputRef = useRef<HTMLSpanElement>(null);
  const runCode = usePython();

  const runCommand = (input: string) => {
    if (input.length === 0) {
      pushCommand({ input, status: "success" });
    } else {
      const commandIndex = commands.length;
      pushCommand({ input, status: "pending" });
      runCode(input, (result) => {
        updateCommand(commandIndex, {
          input,
          ...(result instanceof PyError
            ? { status: "failure", output: result.toString() }
            : { status: "success", output: JSON.stringify(result) }),
        });
      });
    }
  };

  const handleClick = (event: MouseEvent<HTMLElement>) => {
    if (!inputRef.current) return;
    event.preventDefault();

    // If anything is selected, don't focus the input. User will have to click
    // to unselect, and we'll focus the input then.
    const selection = window.getSelection();
    if (selection && selection.toString().length > 0) return;

    inputRef.current?.focus();
  };

  const handleKeyDown = (event: KeyboardEvent<HTMLElement>) => {
    const input = inputRef.current;
    if (!input) return;

    // Cmd+K clears the console on macOS. Let's do the same.
    if (event.key === "k" && (event.ctrlKey || event.metaKey)) {
      event.preventDefault();
      clearVisibleCommands();
      return;
    }

    // Arrow keys for browsing history.
    if (event.key === "ArrowUp") {
      event.preventDefault();
      historyUp();
    } else if (event.key === "ArrowDown") {
      event.preventDefault();
      historyDown();
    }

    // Run the command.
    if (event.key === "Enter") {
      event.preventDefault();
      runCommand(currentInput.trim());
      setCurrentInput("");
    }
  };

  const handleInput = (event: KeyboardEvent<HTMLElement>) => {
    setCurrentInput(event.currentTarget.textContent ?? "");
  };

  // Keep the rendered input in sync with store.
  useEffect(() => {
    const input = inputRef.current;
    if (!input) return;

    // Update the input.
    input.textContent = currentInput;

    // If input is from history, move the cursor to the end.
    if (historyIndex !== null) {
      requestAnimationFrame(() => {
        const range = document.createRange();
        range.selectNodeContents(input);
        range.collapse(false);
        const selection = window.getSelection();
        if (selection) {
          selection.removeAllRanges();
          selection.addRange(range);
        }
      });
    }
  }, [currentInput, historyIndex]);

  // Scroll to bottom on input or command changes.
  // Note: `.scrollIntoView()` is not supported in testing environment.
  useEffect(
    () => inputRef.current?.scrollIntoView?.({ behavior: "smooth" }),
    [currentInput, commands]
  );

  const lineCn = "grid grid-cols-[20px_auto]";
  const textCn = "break-all whitespace-pre-wrap";
  const iconCn = "h-[1lh] flex items-center justify-center";
  return (
    <section
      data-testid="python-shell"
      className={cn(
        "size-full flex flex-col overflow-auto",
        "font-mono select-text bg-gray-900"
      )}
      aria-label="Python Shell"
      onClick={handleClick}
      onKeyDown={handleKeyDown}
    >
      {commands.slice(firstVisibleCommand).map((command, index) => (
        <Fragment key={index}>
          <div className={lineCn}>
            <div className={iconCn}>
              {command.status === "pending" ? (
                <CircleIcon />
              ) : (
                <FilledCircleIcon />
              )}
            </div>
            <pre className={textCn}>{command.input}</pre>
          </div>
          {command.output && (
            <div
              className={cn(
                lineCn,
                command.status === "failure" && "text-red-600"
              )}
            >
              <div className={iconCn}>
                <EnterIcon />
              </div>
              <pre className={textCn}>{command.output}</pre>
            </div>
          )}
        </Fragment>
      ))}
      <div className={lineCn}>
        <div className={iconCn}>
          <DollarIcon />
        </div>
        <span
          className={cn(textCn, "focus:outline-none")}
          aria-label="Command input"
          ref={inputRef}
          contentEditable="plaintext-only"
          spellCheck={false}
          onInput={handleInput}
        />
      </div>
    </section>
  );
};

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
