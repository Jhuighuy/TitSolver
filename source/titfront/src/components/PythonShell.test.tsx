/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ *\
 * Part of BlueTit Solver, licensed under Apache 2.0 with Commons Clause.
 * Commercial use, including SaaS, requires a separate license, see /LICENSE.md
\* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

import "@testing-library/jest-dom/vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { beforeEach, describe, expect, it } from "vitest";

import { PyConnectionProvider } from "~/components/Python";
import { PythonShell } from "~/components/PythonShell";
import { usePythonShellStore } from "~/stores/PythonShellStore";

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

beforeEach(() => usePythonShellStore.getState().reset());

async function setupPythonShell() {
  const result = render(
    <PyConnectionProvider>
      <PythonShell />
    </PyConnectionProvider>
  );
  await waitFor(() => {
    expect(screen.getByTestId("python-shell")).toBeInTheDocument();
  });
  return result;
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

// Note: unfortunately, we cannot test user typing here, `user.keyboard` does
// not work with `contentEditable` elements.
describe("PythonShell", () => {
  it("renders commands", async () => {
    usePythonShellStore.getState().commands = [
      { input: "test1", status: "success", output: "output" },
      { input: "test2", status: "failure", output: "error" },
      { input: "", status: "success" }, // should render no output.
      { input: "test3", status: "pending" },
    ];
    const { asFragment } = await setupPythonShell();
    expect(asFragment()).toMatchSnapshot();
  });

  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  it("clears visible commands", async () => {
    usePythonShellStore.getState().commands = [
      { input: "test", status: "success", output: "output" },
    ];
    await setupPythonShell();

    const user = userEvent.setup();
    await user.click(screen.getByTestId("python-shell"));
    await user.keyboard("{Control>}{k}{/Control}");
    expect(screen.queryByText("success")).not.toBeInTheDocument();
  });

  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  it("browses history up", async () => {
    usePythonShellStore.getState().commands = [
      { input: "test1", status: "success", output: "output" },
      { input: "test2", status: "success", output: "output" },
      { input: "test3", status: "success", output: "output" },
    ];
    usePythonShellStore.getState().clearVisibleCommands();
    await setupPythonShell();

    const user = userEvent.setup();
    await user.click(screen.getByTestId("python-shell"));
    await user.keyboard("{ArrowUp}{ArrowUp}");
    expect(screen.getByText("test2")).toBeInTheDocument();
  });

  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  it("browses history down", async () => {
    usePythonShellStore.getState().commands = [
      { input: "test1", status: "success", output: "output" },
      { input: "test2", status: "success", output: "output" },
      { input: "test3", status: "success", output: "output" },
    ];
    usePythonShellStore.getState().clearVisibleCommands();
    await setupPythonShell();

    const user = userEvent.setup();
    await user.click(screen.getByTestId("python-shell"));
    await user.keyboard("{ArrowUp}{ArrowUp}{ArrowDown}");
    expect(screen.getByText("test3")).toBeInTheDocument();
  });

  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  it("runs commands and displays output", async () => {
    usePythonShellStore.getState().currentInput = "2 + 3";
    usePythonShellStore.getState().clearVisibleCommands();
    await setupPythonShell();

    const user = userEvent.setup();
    await user.click(screen.getByTestId("python-shell"));
    await user.keyboard("{Enter}");
    expect(screen.getByText("2 + 3")).toBeInTheDocument();
    await waitFor(() => {
      expect(screen.getByText("5")).toBeInTheDocument();
    });
  });

  it("does not run empty commands", async () => {
    usePythonShellStore.getState().currentInput = "";
    usePythonShellStore.getState().clearVisibleCommands();
    const { asFragment } = await setupPythonShell();

    const user = userEvent.setup();
    await user.click(screen.getByTestId("python-shell"));
    await user.keyboard("{Enter}");
    expect(asFragment()).toMatchSnapshot();
  });

  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  it("runs commands and displays errors", async () => {
    usePythonShellStore.getState().currentInput = "1 / 0";
    usePythonShellStore.getState().clearVisibleCommands();
    await setupPythonShell();

    const user = userEvent.setup();
    await user.click(screen.getByTestId("python-shell"));
    await user.keyboard("{Enter}");
    expect(screen.getByText("1 / 0")).toBeInTheDocument();
    await waitFor(() => {
      expect(
        screen.getByText("ZeroDivisionError: division by zero")
      ).toBeInTheDocument();
    });
  });
});

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
