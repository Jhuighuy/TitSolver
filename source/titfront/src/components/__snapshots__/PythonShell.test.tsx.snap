// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PythonShell > does not run empty commands 1`] = `
<DocumentFragment>
  <section
    aria-label="Python Shell"
    class="size-full flex flex-col overflow-auto font-mono select-text bg-gray-900"
    data-testid="python-shell"
  >
    <div
      class="grid grid-cols-[20px_auto]"
    >
      <div
        class="h-[1lh] flex items-center justify-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 5c-3.859 0-7 3.141-7 7s3.141 7 7 7 7-3.141 7-7-3.141-7-7-7zm0 12c-2.757 0-5-2.243-5-5s2.243-5 5-5 5 2.243 5 5-2.243 5-5 5z"
          />
          <path
            d="M12 9c-1.627 0-3 1.373-3 3s1.373 3 3 3 3-1.373 3-3-1.373-3-3-3z"
          />
        </svg>
      </div>
      <pre
        class="break-all whitespace-pre-wrap"
      />
    </div>
    <div
      class="grid grid-cols-[20px_auto]"
    >
      <div
        class="h-[1lh] flex items-center justify-center"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="12"
            x2="12"
            y1="1"
            y2="23"
          />
          <path
            d="M17 5H9.5a3.5 3.5 0 0 0 0 7h5a3.5 3.5 0 0 1 0 7H6"
          />
        </svg>
      </div>
      <span
        aria-label="Command input"
        class="break-all whitespace-pre-wrap focus:outline-none"
        contenteditable="plaintext-only"
        spellcheck="false"
      />
    </div>
  </section>
</DocumentFragment>
`;

exports[`PythonShell > renders commands 1`] = `
<DocumentFragment>
  <section
    aria-label="Python Shell"
    class="size-full flex flex-col overflow-auto font-mono select-text bg-gray-900"
    data-testid="python-shell"
  >
    <div
      class="grid grid-cols-[20px_auto]"
    >
      <div
        class="h-[1lh] flex items-center justify-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 5c-3.859 0-7 3.141-7 7s3.141 7 7 7 7-3.141 7-7-3.141-7-7-7zm0 12c-2.757 0-5-2.243-5-5s2.243-5 5-5 5 2.243 5 5-2.243 5-5 5z"
          />
          <path
            d="M12 9c-1.627 0-3 1.373-3 3s1.373 3 3 3 3-1.373 3-3-1.373-3-3-3z"
          />
        </svg>
      </div>
      <pre
        class="break-all whitespace-pre-wrap"
      >
        test1
      </pre>
    </div>
    <div
      class="grid grid-cols-[20px_auto]"
    >
      <div
        class="h-[1lh] flex items-center justify-center"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="15 10 20 15 15 20"
          />
          <path
            d="M4 4v7a4 4 0 0 0 4 4h12"
          />
        </svg>
      </div>
      <pre
        class="break-all whitespace-pre-wrap"
      >
        output
      </pre>
    </div>
    <div
      class="grid grid-cols-[20px_auto]"
    >
      <div
        class="h-[1lh] flex items-center justify-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 5c-3.859 0-7 3.141-7 7s3.141 7 7 7 7-3.141 7-7-3.141-7-7-7zm0 12c-2.757 0-5-2.243-5-5s2.243-5 5-5 5 2.243 5 5-2.243 5-5 5z"
          />
          <path
            d="M12 9c-1.627 0-3 1.373-3 3s1.373 3 3 3 3-1.373 3-3-1.373-3-3-3z"
          />
        </svg>
      </div>
      <pre
        class="break-all whitespace-pre-wrap"
      >
        test2
      </pre>
    </div>
    <div
      class="grid grid-cols-[20px_auto] text-red-600"
    >
      <div
        class="h-[1lh] flex items-center justify-center"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="15 10 20 15 15 20"
          />
          <path
            d="M4 4v7a4 4 0 0 0 4 4h12"
          />
        </svg>
      </div>
      <pre
        class="break-all whitespace-pre-wrap"
      >
        error
      </pre>
    </div>
    <div
      class="grid grid-cols-[20px_auto]"
    >
      <div
        class="h-[1lh] flex items-center justify-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 5c-3.859 0-7 3.141-7 7s3.141 7 7 7 7-3.141 7-7-3.141-7-7-7zm0 12c-2.757 0-5-2.243-5-5s2.243-5 5-5 5 2.243 5 5-2.243 5-5 5z"
          />
          <path
            d="M12 9c-1.627 0-3 1.373-3 3s1.373 3 3 3 3-1.373 3-3-1.373-3-3-3z"
          />
        </svg>
      </div>
      <pre
        class="break-all whitespace-pre-wrap"
      />
    </div>
    <div
      class="grid grid-cols-[20px_auto]"
    >
      <div
        class="h-[1lh] flex items-center justify-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12c0 3.859 3.14 7 7 7 3.859 0 7-3.141 7-7s-3.141-7-7-7c-3.86 0-7 3.141-7 7zm12 0c0 2.757-2.243 5-5 5s-5-2.243-5-5 2.243-5 5-5 5 2.243 5 5z"
          />
        </svg>
      </div>
      <pre
        class="break-all whitespace-pre-wrap"
      >
        test3
      </pre>
    </div>
    <div
      class="grid grid-cols-[20px_auto]"
    >
      <div
        class="h-[1lh] flex items-center justify-center"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="12"
            x2="12"
            y1="1"
            y2="23"
          />
          <path
            d="M17 5H9.5a3.5 3.5 0 0 0 0 7h5a3.5 3.5 0 0 1 0 7H6"
          />
        </svg>
      </div>
      <span
        aria-label="Command input"
        class="break-all whitespace-pre-wrap focus:outline-none"
        contenteditable="plaintext-only"
        spellcheck="false"
      />
    </div>
  </section>
</DocumentFragment>
`;
